# Introduction to Scikit-Learn (sklearn)

This notebook demonstrates some of the most useful functions of the beautiful Scikit-Learn (sklearn) library.

What we are going to cover?:
1. An end-to-end Scikit-Learn workflow.
2. Getting data ready.
3. Choose the right estimator/algorithm for our problems.
4. Fit the model/algorithm and use it to make predictions on our data.
5. Evaluating a model.
6. Improve a model.
7. Save and load a trained model.
8. Putting it all together!

## 1. An end-to-end Scikit-Learn workflow

In [1]:
# Get the data ready
import pandas as pd
heart_disease=pd.read_csv("heart-disease.csv")
heart_disease

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

In [2]:
import numpy as np

In [3]:
# Create X (feature matrix):
X=heart_disease.drop("target", axis=1)

# Create y (labels):
y=heart_disease["target"]

In [4]:
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [5]:
# 2. Choose the right model and hyperparameters
from sklearn.ensemble import RandomForestClassifier
clf=RandomForestClassifier(n_estimators=100)

# We'll keep the default hyperparameters
clf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'sqrt',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'monotonic_cst': None,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

In [6]:
# 3. Fit the model to the training data
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test= train_test_split(X,y, test_size=0.2)

In [7]:
clf.fit(X_train, y_train)

RandomForestClassifier()

In [8]:
# make a prediction:
y_label=clf.predict(np.array([0,2,3,4]))

D:\mlpro\env\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


ValueError: Expected 2D array, got 1D array instead:
array=[0. 2. 3. 4.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

In [9]:
X_test

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
38    65    0   2       155   269    0        1      148      0      0.8   
86    68    1   2       118   277    0        1      151      0      1.0   
74    43    0   2       122   213    0        1      165      0      0.2   
211   61    1   0       120   260    0        1      140      1      3.6   
105   68    0   2       120   211    0        0      115      0      1.5   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
109   50    0   0       110   254    0        0      159      0      0.0   
155   58    0   0       130   197    0        1      131      0      0.6   
16    58    0   2       120   340    0        1      172      0      0.0   
181   65    0   0       150   225    0        0      114      0      1.0   
259   38    1   3       120   231    0        1      182      1      3.8   

     slope  ca  thal  
38       2   0     2  
86       2   1     3  
74       1   0     2  
211      1   1     3  
105      1   0     2  
..     ...  ..   ...  
109      2   0     2  
155      1   0     2  
16       2   0     2  
181      1   3     3  
259      1   0     3  

[61 rows x 13 columns]

In [10]:
y_preds=clf.predict(X_test)

In [11]:
y_preds

array([1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1,
       1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1], dtype=int64)

In [12]:
y_test

38     1
86     1
74     1
211    0
105    1
      ..
109    1
155    1
16     1
181    0
259    0
Name: target, Length: 61, dtype: int64

In [13]:
# 4. Evaluate the model on trainig data and test data:
clf.score(X_train, y_train)

1.0

In [14]:
clf.score(X_test, y_test)

0.7868852459016393

In [15]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
print(classification_report(y_test, y_preds))

              precision    recall  f1-score   support

           0       0.76      0.73      0.75        26
           1       0.81      0.83      0.82        35

    accuracy                           0.79        61
   macro avg       0.78      0.78      0.78        61
weighted avg       0.79      0.79      0.79        61



In [16]:
confusion_matrix(y_test, y_preds)

array([[19,  7],
       [ 6, 29]], dtype=int64)

In [17]:
accuracy_score(y_test, y_preds)

0.7868852459016393

In [18]:
# 5. Improve the model:
# Try different amounts of n_estimators
np.random.seed(42)
for i in range(10,100,10):
    print(f"Trying model with {i} estimators...")
    clf=RandomForestClassifier(n_estimators=i).fit(X_train, y_train)
    print(f"Model accuracy on test set:{clf.score(X_test, y_test)*100:.2f}%")
    print("")

Trying model with 10 estimators...
Model accuracy on test set:81.97%

Trying model with 20 estimators...
Model accuracy on test set:72.13%

Trying model with 30 estimators...
Model accuracy on test set:80.33%

Trying model with 40 estimators...
Model accuracy on test set:77.05%

Trying model with 50 estimators...
Model accuracy on test set:80.33%

Trying model with 60 estimators...
Model accuracy on test set:77.05%

Trying model with 70 estimators...
Model accuracy on test set:77.05%

Trying model with 80 estimators...
Model accuracy on test set:80.33%

Trying model with 90 estimators...
Model accuracy on test set:80.33%



In [19]:
# 6. Save a model and load it
import pickle

pickle.dump(clf, open("random_forest_model_1.pkl", "wb"))

In [20]:
loaded_model=pickle.load(open("random_forest_model_1.pkl", "rb"))
loaded_model.score(X_test, y_test)

0.8032786885245902

# Getting our data ready to be used with machine learning:

Three main things we can do:
1. Split the data into features and labels (Usually `X` & `y`).
2. Filling (also called imputing) or disregarding missing values.
3. Converting non-numerical values to numerical-values (also called feature encoding)

In [21]:
heart_disease

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

In [22]:
X=heart_disease.drop('target', axis=1)

In [23]:
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [24]:
y=heart_disease["target"]
y.head(5)

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

In [25]:
# Split the data into training and test data:
from sklearn.model_selection import train_test_split

In [26]:
X_train, X_test, y_train, y_test= train_test_split(X, y, test_size=0.2)

In [27]:
X_train

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
70    54    1   2       120   258    0        0      147      0      0.4   
177   64    1   2       140   335    0        1      158      0      0.0   
284   61    1   0       140   207    0        0      138      1      1.9   
29    53    1   2       130   197    1        0      152      0      1.2   
71    51    1   2        94   227    0        1      154      1      0.0   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
50    51    0   2       130   256    0        0      149      0      0.5   
229   64    1   2       125   309    0        1      131      1      1.8   
37    54    1   2       150   232    0        0      165      0      1.6   
157   35    1   1       122   192    0        1      174      0      0.0   
237   60    1   0       140   293    0        0      170      0      1.2   

     slope  ca  thal  
70       1   0     3  
177      2   0     2  
284      2   1     3  
29       0   0     2  
71       2   1     3  
..     ...  ..   ...  
50       2   0     2  
229      1   0     3  
37       2   0     3  
157      2   0     2  
237      1   2     3  

[242 rows x 13 columns]

In [28]:
X_test

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
186   60    1   0       130   253    0        1      144      1      1.4   
101   59    1   3       178   270    0        0      145      0      4.2   
131   49    0   1       134   271    0        1      162      0      0.0   
129   74    0   1       120   269    0        0      121      1      0.2   
106   69    1   3       160   234    1        0      131      0      0.1   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
198   62    1   0       120   267    0        1       99      1      1.8   
242   64    1   0       145   212    0        0      132      0      2.0   
24    40    1   3       140   199    0        1      178      1      1.4   
275   52    1   0       125   212    0        1      168      0      1.0   
235   51    1   0       140   299    0        1      173      1      1.6   

     slope  ca  thal  
186      2   1     3  
101      0   0     3  
131      1   0     2  
129      2   1     2  
106      1   1     2  
..     ...  ..   ...  
198      1   2     3  
242      1   2     1  
24       2   0     3  
275      2   2     3  
235      2   0     3  

[61 rows x 13 columns]

In [29]:
y_train

70     1
177    0
284    0
29     1
71     1
      ..
50     1
229    0
37     1
157    1
237    0
Name: target, Length: 242, dtype: int64

In [30]:
y_test

186    0
101    1
131    1
129    1
106    1
      ..
198    0
242    0
24     1
275    0
235    0
Name: target, Length: 61, dtype: int64

In [31]:
X_train.shape, y_train.shape, X_test.shape, y_test.shape

((242, 13), (242,), (61, 13), (61,))

### 1.1 Make sure its all numerical

In [32]:
heart_disease

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

In [33]:
car_sales=pd.read_csv("car-sales-extended.csv")
car_sales

Make Colour  Odometer (KM)  Doors  Price
0     Honda  White          35431      4  15323
1       BMW   Blue         192714      5  19943
2     Honda  White          84714      4  28343
3    Toyota  White         154365      4  13434
4    Nissan   Blue         181577      3  14043
..      ...    ...            ...    ...    ...
995  Toyota  Black          35820      4  32042
996  Nissan  White         155144      3   5716
997  Nissan   Blue          66604      4  31570
998   Honda  White         215883      4   4001
999  Toyota   Blue         248360      4  12732

[1000 rows x 5 columns]

In [34]:
car_sales.head()

Make Colour  Odometer (KM)  Doors  Price
0   Honda  White          35431      4  15323
1     BMW   Blue         192714      5  19943
2   Honda  White          84714      4  28343
3  Toyota  White         154365      4  13434
4  Nissan   Blue         181577      3  14043

In [35]:
len(car_sales)

1000

In [36]:
car_sales.dtypes

Make             object
Colour           object
Odometer (KM)     int64
Doors             int64
Price             int64
dtype: object

In [37]:
car_sales["Doors"].value_counts()

Doors
4    856
5     79
3     65
Name: count, dtype: int64

In [38]:
# Split into X and y
X= car_sales.drop("Price", axis=1)
X

Make Colour  Odometer (KM)  Doors
0     Honda  White          35431      4
1       BMW   Blue         192714      5
2     Honda  White          84714      4
3    Toyota  White         154365      4
4    Nissan   Blue         181577      3
..      ...    ...            ...    ...
995  Toyota  Black          35820      4
996  Nissan  White         155144      3
997  Nissan   Blue          66604      4
998   Honda  White         215883      4
999  Toyota   Blue         248360      4

[1000 rows x 4 columns]

In [39]:
y= car_sales["Price"]
y

0      15323
1      19943
2      28343
3      13434
4      14043
       ...  
995    32042
996     5716
997    31570
998     4001
999    12732
Name: Price, Length: 1000, dtype: int64

In [40]:
# Split into training and test sets
X_train, X_test, y_train, y_test= train_test_split(X, y, test_size=0.2)

In [41]:
# Build a machine learning model
from sklearn.ensemble import RandomForestRegressor
model= RandomForestRegressor()
model.fit(X_train, y_train)
model.score(X_test, y_test)

ValueError: could not convert string to float: 'Toyota'

In [42]:
# Turn categories into numbers
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
categorical_features= ["Make", "Colour", "Doors"]
one_hot= OneHotEncoder()
transformer= ColumnTransformer([('one_hot',
                               one_hot,
                               categorical_features)], 
                               remainder="passthrough")
transformed_X= transformer.fit_transform(X)
transformed_X

array([[0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 3.54310e+04],
       [1.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        1.00000e+00, 1.92714e+05],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 8.47140e+04],
       ...,
       [0.00000e+00, 0.00000e+00, 1.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 6.66040e+04],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 2.15883e+05],
       [0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        0.00000e+00, 2.48360e+05]])

In [43]:
pd.DataFrame(transformed_X)

0    1    2    3    4    5    6    7    8    9    10   11        12
0    0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0   35431.0
1    1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  192714.0
2    0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0   84714.0
3    0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  154365.0
4    0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  181577.0
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...       ...
995  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0   35820.0
996  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  155144.0
997  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0   66604.0
998  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  215883.0
999  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  248360.0

[1000 rows x 13 columns]

In [44]:
dummies= pd.get_dummies(car_sales[["Make","Colour","Doors"]])
dummies

Doors  Make_BMW  Make_Honda  Make_Nissan  Make_Toyota  Colour_Black  \
0        4     False        True        False        False         False   
1        5      True       False        False        False         False   
2        4     False        True        False        False         False   
3        4     False       False        False         True         False   
4        3     False       False         True        False         False   
..     ...       ...         ...          ...          ...           ...   
995      4     False       False        False         True          True   
996      3     False       False         True        False         False   
997      4     False       False         True        False         False   
998      4     False        True        False        False         False   
999      4     False       False        False         True         False   

     Colour_Blue  Colour_Green  Colour_Red  Colour_White  
0          False         False       False          True  
1           True         False       False         False  
2          False         False       False          True  
3          False         False       False          True  
4           True         False       False         False  
..           ...           ...         ...           ...  
995        False         False       False         False  
996        False         False       False          True  
997         True         False       False         False  
998        False         False       False          True  
999         True         False       False         False  

[1000 rows x 10 columns]

In [45]:
# Let's refit the model
np.random.seed(42)
X_train, X_test, y_train, y_test= train_test_split(transformed_X, y, test_size=0.2)
model.fit(X_train, y_train)

RandomForestRegressor()

In [46]:
model.score(X_test, y_test)

0.3235867221569877

In [47]:
import sklearn
print('The scikit-learn version is {}.'.format(sklearn.__version__))

The scikit-learn version is 1.5.1.


## 1.2 What if there were missing values?

1. Fill them with some value (also known as imputation).
2. Remove the samples with missing data altogether.

In [48]:
# Import the car sales missing data
car_sales_missing= pd.read_csv("car-sales-extended-missing-data.csv")
car_sales_missing.head()

Make Colour  Odometer (KM)  Doors    Price
0   Honda  White        35431.0    4.0  15323.0
1     BMW   Blue       192714.0    5.0  19943.0
2   Honda  White        84714.0    4.0  28343.0
3  Toyota  White       154365.0    4.0  13434.0
4  Nissan   Blue       181577.0    3.0  14043.0

In [49]:
car_sales_missing.isna().sum()

Make             49
Colour           50
Odometer (KM)    50
Doors            50
Price            50
dtype: int64

In [50]:
# Creating X and y:
X= car_sales_missing.drop("Price", axis=1)
y=car_sales_missing["Price"]

In [51]:
# Let's try and convert our data into numbers:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
categorical_features= ["Make", "Colour", "Doors"]
one_hot=OneHotEncoder()
transformer=ColumnTransformer([('one_hot',
                               one_hot,
                               categorical_features)],
                             remainder="passthrough")
transformed_X=transformer.fit_transform(X)
transformed_X

<1000x16 sparse matrix of type '<class 'numpy.float64'>'
	with 4000 stored elements in Compressed Sparse Row format>

#### Optional: Fill missing data (In new scikit learn, the NaN/None values gives no error)

In [52]:
# Fill the "Make" column
car_sales_missing['Make'].fillna("missing")

# Fill the "Colour" column
car_sales_missing['Colour'].fillna("missing")

# Fill the "Odometer (KM)" column
car_sales_missing['Odometer (KM)'].fillna(car_sales_missing['Odometer (KM)'].mean(numeric_only=True))

# Fill the "Doors" column
car_sales_missing['Doors'].fillna(4)

0      4.0
1      5.0
2      4.0
3      4.0
4      3.0
      ... 
995    4.0
996    3.0
997    4.0
998    4.0
999    4.0
Name: Doors, Length: 1000, dtype: float64

In [53]:
car_sales_missing['Make'].isna()

0      False
1      False
2      False
3      False
4      False
       ...  
995    False
996     True
997    False
998    False
999    False
Name: Make, Length: 1000, dtype: bool

In [54]:
car_sales_missing.isna().sum()

Make             49
Colour           50
Odometer (KM)    50
Doors            50
Price            50
dtype: int64

In [55]:
# Remove rows with missing Price values
car_sales_missing.dropna(inplace=True)

In [56]:
car_sales_missing.isna().sum()

Make             0
Colour           0
Odometer (KM)    0
Doors            0
Price            0
dtype: int64

In [57]:
# Splitting data
X= car_sales_missing.drop('Price', axis=1)
y=car_sales_missing['Price']

In [58]:
# Again doing the same shit
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
categorical_features=['Make', 'Colour', 'Doors']
one_hot=OneHotEncoder()
transformer=ColumnTransformer([('one_hot',
                               one_hot,
                               categorical_features)],
                             remainder="passthrough")
transformed_X=transformer.fit_transform(car_sales_missing)
transformed_X

array([[0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        3.54310e+04, 1.53230e+04],
       [1.00000e+00, 0.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        1.92714e+05, 1.99430e+04],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        8.47140e+04, 2.83430e+04],
       ...,
       [0.00000e+00, 0.00000e+00, 1.00000e+00, ..., 0.00000e+00,
        6.66040e+04, 3.15700e+04],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        2.15883e+05, 4.00100e+03],
       [0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        2.48360e+05, 1.27320e+04]])

In [59]:
transformed_X

array([[0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        3.54310e+04, 1.53230e+04],
       [1.00000e+00, 0.00000e+00, 0.00000e+00, ..., 1.00000e+00,
        1.92714e+05, 1.99430e+04],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        8.47140e+04, 2.83430e+04],
       ...,
       [0.00000e+00, 0.00000e+00, 1.00000e+00, ..., 0.00000e+00,
        6.66040e+04, 3.15700e+04],
       [0.00000e+00, 1.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        2.15883e+05, 4.00100e+03],
       [0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        2.48360e+05, 1.27320e+04]])

### Filling missing data with Scikit-Learn

In [60]:
car_sale_missing=pd.read_csv("car-sales-extended-missing-data.csv")

In [61]:
car_sales

Make Colour  Odometer (KM)  Doors  Price
0     Honda  White          35431      4  15323
1       BMW   Blue         192714      5  19943
2     Honda  White          84714      4  28343
3    Toyota  White         154365      4  13434
4    Nissan   Blue         181577      3  14043
..      ...    ...            ...    ...    ...
995  Toyota  Black          35820      4  32042
996  Nissan  White         155144      3   5716
997  Nissan   Blue          66604      4  31570
998   Honda  White         215883      4   4001
999  Toyota   Blue         248360      4  12732

[1000 rows x 5 columns]

In [62]:
car_sale_missing

Make Colour  Odometer (KM)  Doors    Price
0     Honda  White        35431.0    4.0  15323.0
1       BMW   Blue       192714.0    5.0  19943.0
2     Honda  White        84714.0    4.0  28343.0
3    Toyota  White       154365.0    4.0  13434.0
4    Nissan   Blue       181577.0    3.0  14043.0
..      ...    ...            ...    ...      ...
995  Toyota  Black        35820.0    4.0  32042.0
996     NaN  White       155144.0    3.0   5716.0
997  Nissan   Blue        66604.0    4.0  31570.0
998   Honda  White       215883.0    4.0   4001.0
999  Toyota   Blue       248360.0    4.0  12732.0

[1000 rows x 5 columns]

In [63]:
car_sale_missing.isna().sum()

Make             49
Colour           50
Odometer (KM)    50
Doors            50
Price            50
dtype: int64

In [64]:
# Dropping the Price value rows which has missing values:
car_sale_missing.dropna(subset=['Price'], inplace=True)
car_sale_missing.isna().sum()

Make             47
Colour           46
Odometer (KM)    48
Doors            47
Price             0
dtype: int64

In [65]:
# Split into X and y:
X= car_sale_missing.drop("Price", axis=1)
y=car_sale_missing["Price"]

In [66]:
# Fill missing values with Scikit-learn:
from sklearn.impute import SimpleImputer # Filling missing values is called imputation.
from sklearn.compose import ColumnTransformer

# Fill categorical values 'missing' and numerical values with mean:
cat_imputer= SimpleImputer(strategy="constant", fill_value="missing")
door_imputer= SimpleImputer(strategy="constant", fill_value=4)
num_imputer= SimpleImputer(strategy="mean")

# Define columns:
cat_feature=["Make", "Colour"]
door_feature=["Doors"]
num_feature=["Odometer (KM)"]

# Create an imputer (something that fills missing values):
imputer= ColumnTransformer([("cat_imputer", cat_imputer, cat_feature),
                          ("door_imputer", door_imputer, door_feature),
                          ("num_imputer", num_imputer, num_feature)])

# Transform the data:
filled_X= imputer.fit_transform(X)
filled_X

array([['Honda', 'White', 4.0, 35431.0],
       ['BMW', 'Blue', 5.0, 192714.0],
       ['Honda', 'White', 4.0, 84714.0],
       ...,
       ['Nissan', 'Blue', 4.0, 66604.0],
       ['Honda', 'White', 4.0, 215883.0],
       ['Toyota', 'Blue', 4.0, 248360.0]], dtype=object)

In [67]:
car_sales_filled= pd.DataFrame(filled_X, columns=["Make","Colour","Door","Odometer (KM)"])
car_sales_filled.head()

Make Colour Door Odometer (KM)
0   Honda  White  4.0       35431.0
1     BMW   Blue  5.0      192714.0
2   Honda  White  4.0       84714.0
3  Toyota  White  4.0      154365.0
4  Nissan   Blue  3.0      181577.0

In [68]:
car_sales_filled.isna().sum()

Make             0
Colour           0
Door             0
Odometer (KM)    0
dtype: int64

In [69]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
categorical_features=["Make","Colour","Door"]
one_hot=OneHotEncoder()
transformer=ColumnTransformer([("one_hot",
                               one_hot,
                               categorical_features)])

transformed_X= transformer.fit_transform(car_sales_filled)
transformed_X

<950x14 sparse matrix of type '<class 'numpy.float64'>'
	with 2850 stored elements in Compressed Sparse Row format>

In [70]:
# Now we've got our data as numbers and filled (no missing values), Lets fit a model:
np.random.seed(42)
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test= train_test_split(transformed_X, y, test_size=0.2)
model=RandomForestRegressor()
model.fit(X_train, y_train)
model.score(X_test, y_test)

0.09545688096364557

In [71]:
# Check length of transformed data (filled and one-hot encoded) vs the length of of original one:
len(car_sales_filled), len(car_sales)

(950, 1000)

#### Note:
We dropped columns which has no values in Price column and results in reduction of 50 rows.

## 2. Choose the right estimator/algorithm for your problem:

Some things to note:
* Sklearn refers to machine learning models, algorithms as estimators.
* Classification problems- predicting a category (heart disease or not).
    * Sometimes you'll see a `clf` (short for classifier) used as a classfication estimator.
* Regression problems- Predicting a number (selling price of a car)
<br>
If you are working on a machine learning problem and looking to use Sklearn and not sure what model you should use, refer to the sklearn machine learning map: <a href="https://scikit-learn.org/1.5/machine_learning_map.html">Choosing right estimator</a>

### 2.1 Picking a machine learnig model for a regression problem:

Let's use California Housing Dataset: <a href="https://scikit-learn.org/1.5/modules/generated/sklearn.datasets.fetch_california_housing.html">Here is the info</a>

In [72]:
# Get California Housing dataset:
from sklearn.datasets import fetch_california_housing
housing = fetch_california_housing()
housing

{'data': array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
           37.88      , -122.23      ],
        [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
           37.86      , -122.22      ],
        [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
           37.85      , -122.24      ],
        ...,
        [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
           39.43      , -121.22      ],
        [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
           39.43      , -121.32      ],
        [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
           39.37      , -121.24      ]]),
 'target': array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894]),
 'frame': None,
 'target_names': ['MedHouseVal'],
 'feature_names': ['MedInc',
  'HouseAge',
  'AveRooms',
  'AveBedrms',
  'Population',
  'AveOccup',
  'Latitude',
  'Longitude'],
 'DESCR': '.. _california_housing_dataset:\n

In [73]:
housing_df= pd.DataFrame(housing["data"], columns=housing["feature_names"])

In [74]:
housing_df

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  
0        -122.23  
1        -122.22  
2        -122.24  
3        -122.25  
4        -122.25  
...          ...  
20635    -121.09  
20636    -121.21  
20637    -121.22  
20638    -121.32  
20639    -121.24  

[20640 rows x 8 columns]

In [75]:
housing_df["target"] = housing["target"]
housing_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422

In [76]:
housing_df = housing_df.drop("MedHouseVal", axis=1)
housing_df

KeyError: "['MedHouseVal'] not found in axis"

In [80]:
# Import algorithm/estimator
from sklearn.linear_model import Ridge

# Setup random seed
np.random.seed(42)

# Create the data
X=housing_df.drop("target", axis=1)
y=housing_df["target"] # median house price in $100,000

# Split into train and test dataset
X_train, X_test, y_train, y_test= train_test_split(X, y, test_size=0.2)

# Instantiate and fit the model (on the training set):
model=Ridge()
model.fit(X_train, y_train)

# Check the score of the model (on the test dataset)
model.score(X_test, y_test)

0.5758549611440128

What if `Ridge` didn't work or the score didn't fit our needs? <br>
Well, we could always try a different model...

How about we try an ensemble model (an ensemble model is a combination of smaller models to try and make better predictions than just a single model). <br>
Sklearn ensemble model: <a href="https://scikit-learn.org/stable/modules/ensemble.html">Scikit-Learn Ensemble Model</a>

In [81]:
# Import the RandomForestRegressor model class from the ensemble modlue
from sklearn.ensemble import RandomForestRegressor

# Setup random seed
np.random.seed(42)

# Create the data
X= housing_df.drop("target", axis=1)
y=housing_df["target"]

# Splitting our data
X_train, X_test, y_train, y_test= train_test_split(X,y,test_size=0.2)

# Create the random forest model
model=RandomForestRegressor()
model.fit(X_train, y_train)

# Check the score of the model:
model.score(X_test, y_test)

0.806652667101436

## 2.2 Choosing an estimator for a classification problem:
Let's go to the map... : <a href = "https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html"> Scikit-Learn Map </a>

In [82]:
heart_disease=pd.read_csv("heart-disease.csv")
heart_disease

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

Consulting the map and it says to try `LinearSVC`.

In [83]:
# Import the LineSVC estimator class
from sklearn.svm import LinearSVC

# Setup random seed
np.random.seed(42)

# Make the data ready
X=heart_disease.drop("target", axis=1)
y=heart_disease["target"]

# Split the data into train and test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Get the model ready
clf=LinearSVC()
clf.fit(X_train, y_train)

# Check the score of the model on the test dataset
clf.score(X_test, y_test)

0.8688524590163934

We got a good value of `0.8688524590163934` on running our model, but Daniel didn't. So he used RandomForestClassifier to get better results.

In [84]:
# Import the RandomForestClassifier class
from sklearn.ensemble import RandomForestClassifier

# Setupt random seed
np.random.seed(42)

# Make the data ready
X = heart_disease.drop("target", axis=1)
y = heart_disease["target"]

# Split the data into train and test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y ,test_size=0.2)

# Fitting the data into our model
clf = RandomForestClassifier()
clf.fit(X_train, y_train)

# Checking the score of our model on test data:
clf.score(X_test, y_test)

0.8524590163934426

As you can see we also got a decent result using RandomForestClassifier (A score of `0.8524590163934426` for the first time)

Tidbit:
1. If you have structured data, used ensemble methods
2. If you have unstructured data, use deep learning or transfer learning

## 3. Fit the model/algorithm on our data and use it to make predictions

### 3.1 Fitting the model to the data

Different names for:
* `X` = features, features variables, data
* `y` = labels, targets, target variables

In [85]:
# Import the RandomForestClassifier class
from sklearn.ensemble import RandomForestClassifier

# Setupt random seed
np.random.seed(42)

# Make the data ready
X = heart_disease.drop("target", axis=1)
y = heart_disease["target"]

# Split the data into train and test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y ,test_size=0.2)

# Fitting the data into our model (training the machine learning model)
clf = RandomForestClassifier()
clf.fit(X_train, y_train)

# Checking the score of our model on test data (use the patterns the model has learned already):
clf.score(X_test, y_test)

0.8524590163934426

In [86]:
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [87]:
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

In [88]:
y.tail()

298    0
299    0
300    0
301    0
302    0
Name: target, dtype: int64

### 3.2 Make predictions using a machine learning model
There are two ways to make predictions:
1. `predict()`
2. `predict_proba()`

In [89]:
# Use a trained model to make predictions:
clf.predict(np.array([1,7,8,3,4])) # This doesn't work

D:\mlpro\env\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


ValueError: Expected 2D array, got 1D array instead:
array=[1. 7. 8. 3. 4.].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

In [90]:
X_test.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
179   57    1   0       150   276    0        0      112      1      0.6   
228   59    1   3       170   288    0        0      159      0      0.2   
111   57    1   2       150   126    1        1      173      0      0.2   
246   56    0   0       134   409    0        0      150      1      1.9   
60    71    0   2       110   265    1        0      130      0      0.0   

     slope  ca  thal  
179      1   1     1  
228      1   0     3  
111      2   1     3  
246      1   2     3  
60       2   1     2

In [91]:
clf.predict(X_test)

array([0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0], dtype=int64)

In [92]:
np.array([y_test])

array([[0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0,
        0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0]], dtype=int64)

In [93]:
# Compare predictions to truth labels to evaluate the model
y_preds = clf.predict(X_test)
np.mean(y_preds == y_test)

0.8524590163934426

In [94]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_preds)

0.8524590163934426

Make predictiond with `predict_proba()`

In [95]:
# predict_proba returns probabilities of a classficationm label:
clf.predict_proba(X_test[:5])

array([[0.89, 0.11],
       [0.49, 0.51],
       [0.43, 0.57],
       [0.84, 0.16],
       [0.18, 0.82]])

In [96]:
# Let's predict() on the same data...
clf.predict(X_test[:5])

array([0, 1, 1, 0, 1], dtype=int64)

`predict_proba()` is taking up the rows and converting each 0 (False), and 1 (True) into a an array (as you can see in [0.89, 0.11]). 0.89 is the probability of target to be 0 (False). Whereas 0.11 is the probability of target being 1 (True). If you sum 0.89+0.11 you will get 1 which is the highest probability a value can get.

`predict()` can also be used for regression modeels.

In [97]:
housing_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422

In [98]:
from sklearn.ensemble import RandomForestRegressor

np.random.seed(42)

# Create the data
X = housing_df.drop("target", axis=1)
y = housing_df["target"]

# Split into training and sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Create model instance
model = RandomForestRegressor()

# Fit the model
model.fit(X_train, y_train)

RandomForestRegressor()

In [99]:
# Making predictions
y_preds = model.predict(X_test)
y_preds[:10]

array([0.4939   , 0.75494  , 4.9285964, 2.54024  , 2.33176  , 1.6602201,
       2.3431   , 1.66311  , 2.47489  , 4.8344779])

In [100]:
y_preds[:10]

array([0.4939   , 0.75494  , 4.9285964, 2.54024  , 2.33176  , 1.6602201,
       2.3431   , 1.66311  , 2.47489  , 4.8344779])

In [101]:
np.array(y_test[:10])

array([0.477  , 0.458  , 5.00001, 2.186  , 2.78   , 1.587  , 1.982  ,
       1.575  , 3.4    , 4.466  ])

In [102]:
# Compare the predictions to the truth
from sklearn.metrics import mean_absolute_error
mean_absolute_error(y_test, y_preds)

0.32656738464147306

## 4. Evaluating a machine learning model

Three ways to evaluate Scikit-Learn models/estimators:

1. Estimator's built-in `score()` method.
2. The `scoring` parameter.
3. Problem-specific metric functions.

You can read more about these here: <a href="https://scikit-learn.org/stable/modules/model_evaluation.html">Scikit-Learn Model Evaluation</a>

### 4.1 Evaluating a model with the `score` method

In [103]:
from sklearn.ensemble import RandomForestClassifier

np.random.seed(42)

# Create X & y
X = heart_disease.drop("target", axis=1)
y = heart_disease["target"]

# Create train/tests
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Create classifier model instance:
clf = RandomForestClassifier()

# Fit classifier to training data:
clf.fit(X_train, y_train)

RandomForestClassifier()

In [104]:
# The highest value for the .score() method is 1.0, the lowest is 0.0
clf.score(X_train, y_train)

1.0

In [105]:
clf.score(X_test, y_test)

0.8524590163934426

Let's use the `score()` on our regression problem...

In [106]:
from sklearn.ensemble import RandomForestRegressor

np.random.seed(42)

X = housing_df.drop("target", axis=1)
y = housing_df["target"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

model = RandomForestRegressor(n_estimators=1000)

model.fit(X_train, y_train)

RandomForestRegressor(n_estimators=1000)

In [107]:
model.score(X_train, y_train)

0.9746617116821515

In [108]:
model.score(X_test, y_test)

0.8083315928399709

In [109]:
model.score(X_test, y_test)

0.8083315928399709

## 4.2 Evaluating a model using the `scoring` parameter

In [110]:
from sklearn.model_selection import cross_val_score

from sklearn.ensemble import RandomForestClassifier

np.random.seed(42)

X = heart_disease.drop("target", axis=1)
y = heart_disease["target"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

clf = RandomForestClassifier()

clf.fit(X_train, y_train)

RandomForestClassifier()

In [111]:
clf.score(X_test, y_test)

0.8524590163934426

In [112]:
cross_val_score(clf, X, y, cv=5)

array([0.81967213, 0.86885246, 0.81967213, 0.78333333, 0.76666667])

In [113]:
cross_val_score(clf, X, y)

array([0.83606557, 0.8852459 , 0.7704918 , 0.8       , 0.8       ])

In [114]:
np.random.seed(42)

# Single training and test split
clf_single_score = clf.score(X_test, y_test)

# Take the mean of 5-fold cross-validation score
clf_cross_val_score = np.mean(cross_val_score(clf, X, y))

# Compare the two
clf_single_score, clf_cross_val_score

(0.8524590163934426, 0.8248087431693989)

### 4.2.1 Classification model evaluation metrics
1. Accuracy
2. Area under ROC curve
3. Confusion matrix
4. CLassification report

**Accuracy**

In [115]:
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier

np.random.seed(42)

X = heart_disease.drop("target", axis=1)
y = heart_disease["target"]

clf = RandomForestClassifier()
cross_val_score = cross_val_score(clf, X, y)

In [116]:
np.mean(cross_val_score)

0.8248087431693989

In [117]:
print(f"Heart Disease Classifier Cross-Validated Accuracy: {np.mean(cross_val_score)*100:.2f}%")

Heart Disease Classifier Cross-Validated Accuracy: 82.48%


**Area under Reciever Operating Characterstic (AUC/ROC) Curve**
* Area under curve (AUC)
* ROC curve

ROC curves are a comparision of a model's true positive rate (tpr) versus a models false positive rate (fpr).

* True positive = model predicts 1 when truth is 0.
* False positive = model predicts 1 when truth is 0.
* True negative = model predicts 0 when truth is 0.
* False negative = model predicts 0 when truth is 1.

In [119]:
from sklearn.metrics import roc_curve

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Fit the classifier:
clf.fit(X_train, y_train)

# Make the predictions with probabilities
y_probs = clf.predict_proba(X_test)

y_probs[:10], len(y_probs)

(array([[0.7 , 0.3 ],
        [0.59, 0.41],
        [0.92, 0.08],
        [0.08, 0.92],
        [0.07, 0.93],
        [0.03, 0.97],
        [0.27, 0.73],
        [0.91, 0.09],
        [0.38, 0.62],
        [0.1 , 0.9 ]]),
 61)

In [121]:
y_probs_positive = y_probs[:, 1]
y_probs_positive[:10]

array([0.3 , 0.41, 0.08, 0.92, 0.93, 0.97, 0.73, 0.09, 0.62, 0.9 ])

In [122]:
# Calculate fpr, tpr and threshold
fpr, tpr, threshold = roc_curve(y_test, y_probs_positive)

# Check the false positive rates
fpr

array([0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.03703704, 0.03703704, 0.07407407, 0.07407407, 0.11111111,
       0.11111111, 0.14814815, 0.14814815, 0.18518519, 0.22222222,
       0.2962963 , 0.2962963 , 0.33333333, 0.37037037, 0.48148148,
       0.51851852, 0.59259259, 0.7037037 , 0.81481481, 0.85185185,
       0.92592593, 1.        ])

In [123]:
# Create a function for plotting ROC curves
import matplotlib.pyplot as plt

def plot_roc_curve(fpr, tpr):
    """Plots a ROC curve given the false positive rate (fpr) and true positive rate (tpr"""

SyntaxError: incomplete input (2787477380.py, line 4)